apollo-federation/src/schema/schema_upgrader.rs

```diff
@@ -960,9 +961,77 @@ pub fn upgrade_subgraphs_if_necessary(
             ),
         }]);
     }
+
     Ok(upgraded_subgraphs)
 }
 
+/// Upgrade subgraphs if necessary and validate the result.
+/// - Fed 2 input subgraphs are not upgraded.
+/// - Also, normalizes root types if necessary.
+/// - Unchanged subgraphs are returned as-is.
+// PORT_NOTE: In JS, this returns upgraded subgraphs along with a set of messages about what changed.
+// However, those messages were never used, so we have omitted them here.
+#[instrument(skip(subgraphs))]
+pub fn upgrade_subgraphs_if_necessary(
+    subgraphs: Vec<Subgraph<Expanded>>,
+) -> Result<Vec<Subgraph<Validated>>, Vec<CompositionError>> {
+    // Upgrade subgraphs (if necessary)
+    let upgraded = inner_upgrade_subgraphs_if_necessary(subgraphs)?;
+
+    let mut errors: Vec<CompositionError> = vec![];
+
+    // Normalize root types (if necessary)
+    let normalized: Vec<_> = upgraded
+        .into_iter()
+        .filter_map(|subgraph| match subgraph {
+            Either::Left(s) => match s.normalize_root_types() {
+                Ok(s) => Some(s),
```

Comment:
This should be
```suggestion
                Ok(s) => Some(Either::Left(s)),
```

On line 1011, you are trying to match against `Either::Left(s)` but you aren't emitting that here.

Reply:
Gotcha 👍🏾